libraries/chain/db_update.cpp

```diff
@@ -107,32 +107,19 @@ void database::update_last_irreversible_block()
    const global_property_object& gpo = get_global_properties();
    const dynamic_global_property_object& dpo = get_dynamic_global_properties();
 
-   // TODO for better performance, move this to db_maint, because only need to do it once per maintenance interval
-   vector< const witness_object* > wit_objs;
-   wit_objs.reserve( gpo.active_witnesses.size() );
-   for( const witness_id_type& wid : gpo.active_witnesses )
-      wit_objs.push_back( &(wid(*this)) );
-
    static_assert( GRAPHENE_IRREVERSIBLE_THRESHOLD > 0, "irreversible threshold must be nonzero" );
 
-   // 1 1 1 2 2 2 2 2 2 2 -> 2     .3*10 = 3
-   // 1 1 1 1 1 1 1 2 2 2 -> 1
-   // 3 3 3 3 3 3 3 3 3 3 -> 3
-   // 3 3 3 4 4 4 4 4 4 4 -> 4
-
-   size_t offset = ((GRAPHENE_100_PERCENT - GRAPHENE_IRREVERSIBLE_THRESHOLD) * wit_objs.size() / GRAPHENE_100_PERCENT);
-
-   std::nth_element( wit_objs.begin(), wit_objs.begin() + offset, wit_objs.end(),
-      []( const witness_object* a, const witness_object* b )
-      {
-         return a->last_confirmed_block_num < b->last_confirmed_block_num;
-      } );
-
-   uint32_t new_last_irreversible_block_num = wit_objs[offset]->last_confirmed_block_num;
+   const size_t offset = (GRAPHENE_100_PERCENT - GRAPHENE_IRREVERSIBLE_THRESHOLD)
+                         * gpo.active_witnesses.size() / GRAPHENE_100_PERCENT;
+   const auto& witness_idx = get_index_type<witness_index>().indices().get<by_last_block>();
+   auto itr = witness_idx.end();
+   for( uint32_t i = gpo.active_witnesses.size() - offset; i > 0; --i )
+      --itr;
```

Comment:
I think this is the main place where the performance got (slightly) lost, because we walk `number_of_active_witnesses/2` steps on each new block. Each step costs `log(number_of_all_witnesses)`.

My thoughts:
* maintain a small set containing the witnesses that produced a block after LIB and are still in the active witness list;
* maintain a pointer/iterator to the current witness that holds the LIB, and probably another pointer/iterator to its next witness in the set for better performance;
* when a new block comes,
  * if its witness is already in the set, LIB does not change;
  * otherwise, add the new witness into the set,
    * if the new set size is smaller than or equal to `number_of_active_witnesses/2`, LIB does not change;
    * otherwise, move the pointer/iterator and LIB to the next witness in the set, and remove the old one from the set.

With this approach we walk one, two or three steps on each new block within a smaller set with cost `log(number_of_active_witnesses)` (one for the lookup (if a witness is already in the set), one for adding it, another for removing the old). This can probably be done with a secondary index to keep data consistency. Not sure how much performance gain it will bring though.